File: windows/src/Slipstream.Core/Net/LanGuard.cs
using System.Net;
using System.Net.Sockets;

namespace Slipstream.Core.Net;

public sealed class NonLocalAddressException(IPAddress address)
    : Exception($"Refused non-local address {address}. Slipstream never leaves the local network.")
{
    public IPAddress Address { get; } = address;
}

/// <summary>
/// Spec §11 layer 2: the only addresses Slipstream will connect to or accept from.
/// Applied to both inbound and outbound connections.
/// </summary>
public static class LanGuard
{
    public static bool IsLocal(IPAddress address)
    {
        if (address.IsIPv4MappedToIPv6)
            address = address.MapToIPv4();

        return address.AddressFamily switch
        {
            AddressFamily.InterNetwork => IsLocalV4(address),
            AddressFamily.InterNetworkV6 => IsLocalV6(address),
            _ => false,
        };
    }

    private static bool IsLocalV4(IPAddress address)
    {
        Span<byte> b = stackalloc byte[4];
        if (!address.TryWriteBytes(b, out _)) return false;

        if (b[0] == 10) return true;                          // 10.0.0.0/8
        if (b[0] == 172 && b[1] >= 16 && b[1] <= 31) return true; // 172.16.0.0/12
        if (b[0] == 192 && b[1] == 168) return true;          // 192.168.0.0/16
        if (b[0] == 169 && b[1] == 254) return true;          // 169.254.0.0/16
        if (b[0] == 127) return true;                         // loopback, for tests
        return false;
    }

    private static bool IsLocalV6(IPAddress address)
    {
        if (IPAddress.IsLoopback(address)) return true;
        if (address.IsIPv6LinkLocal) return true;             // fe80::/10

        Span<byte> b = stackalloc byte[16];
        if (!address.TryWriteBytes(b, out _)) return false;
        return (b[0] & 0xFE) == 0xFC;                         // fc00::/7 unique-local
    }

    public static void EnsureLocal(IPAddress address)
    {
        if (!IsLocal(address)) throw new NonLocalAddressException(address);
    }
}

File: android/core/src/main/kotlin/com/slipstream/core/net/LanGuard.kt
package com.slipstream.core.net

import java.net.Inet4Address
import java.net.Inet6Address
import java.net.InetAddress

class NonLocalAddressException(val address: InetAddress) :
    Exception("Refused non-local address $address. Slipstream never leaves the local network.")

/**
 * Spec §11 layer 2: the only addresses Slipstream will connect to or accept from.
 * Applied to inbound and outbound connections alike. Must agree byte-for-byte with
 * the C# LanGuard, or the two ends disagree about whether they are allowed to talk.
 */
object LanGuard {
    fun isLocal(address: InetAddress): Boolean = when (address) {
        is Inet4Address -> isLocalV4(address.address)
        is Inet6Address -> isLocalV6(address)
        else -> false
    }

    private fun isLocalV4(b: ByteArray): Boolean {
        val o0 = b[0].toInt() and 0xFF
        val o1 = b[1].toInt() and 0xFF
        return when {
            o0 == 10 -> true                       // 10.0.0.0/8
            o0 == 172 && o1 in 16..31 -> true      // 172.16.0.0/12
            o0 == 192 && o1 == 168 -> true         // 192.168.0.0/16
            o0 == 169 && o1 == 254 -> true         // link-local
            o0 == 127 -> true                      // loopback, for tests
            else -> false
        }
    }

    private fun isLocalV6(address: Inet6Address): Boolean {
        if (address.isLoopbackAddress || address.isLinkLocalAddress) return true
        return (address.address[0].toInt() and 0xFE) == 0xFC   // fc00::/7 ULA
    }

    fun ensureLocal(address: InetAddress) {
        if (!isLocal(address)) throw NonLocalAddressException(address)
    }
}
